racket/collects/racket/hash-code.rkt

```diff
@@ -0,0 +1,46 @@
+#lang racket/base
+
+(provide mix-hash-code hash-code-combine hash-code-combine-unordered)
+
+(require racket/fixnum)
+
+;; unsigned integer view of -2, not a fixnum
+(define (um2) (arithmetic-shift (most-positive-fixnum) 1))
+
+;; fixnum shift-left logical with wraparound
+(define fxsll/wraparound fxlshift/wraparound)
+
+;; fixnum logical exclusive or
+(define fxlogxor fxxor)
+
+;; fixnum shift-right logical
+;; shifting negative fixnums can produce large positive fixnums
+(define (fxsrl a b)
+  (cond
+    [(and (negative? a) (positive? b))
+     (arithmetic-shift (bitwise-and a (um2)) (- b))]
+    [else (fxrshift a b)]))
+
+;; Adapted from racket/src/cs/rumble/hash-code.ss
+;; which is adapted from ChezScheme/s/newhash.ss
+
+;; maps non-fixnum integers to positive fixnums
+(define (->fx v)
+  (if (fixnum? v)
+      v
+      (bitwise-and v (most-positive-fixnum))))
+
+(define (mix-hash-code hc)
+  (fxmix-hash-code (->fx hc)))
+
+(define (fxmix-hash-code hc)
+  (let ([hc2 (fx+/wraparound hc (fxsll/wraparound (fx+/wraparound hc 1) 10))])
+    (fxlogxor hc2 (fxsrl hc2 6))))
+
+(define (hash-code-combine hc . rst)
+  (for/fold ([acc (->fx hc)]) ([v (in-list rst)])
+    (fx+/wraparound (fxmix-hash-code acc) (->fx v))))
```

Comment:
Problem: if `(hash-code-combine A B)` = `(hash-code-combine-unordered (mix-hash-code A) B)`, then `(hash-code-combine-unordered (hash-code-combine A B) C)` = `(hash-code-combine-unordered (hash-code-combine A C) B)`, which might cause hash collisions where users wouldn't expect, for example `{[A, B], C}` might collide with `{[A, C], B}`.

If instead `(hash-code-combine A B)` is made to equal `(mix-hash-code (hash-code-combine-unordered (mix-hash-code A) B)`, then `{[A, B], C}` and `{[A, C], B}` implemented this way would be different.